### PR TITLE
crimson/object_context: make obc interruptible

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/object_context.h [hunk 1]
```diff
@@ -84,6 +85,14 @@ class ObjectContext : public Blocker,
     loaded = true;
   }
 
+  /// pass the provided exception to any waiting consumers of this ObjectContext
+  void interrupt(std::exception_ptr ex) {
```
Comment (tchaikov) [inline src/crimson/osd/object_context.h:89]: better off passing a copyable exception.

**Comment by tchaikov** [comment]
this change is based on #37102

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

File: src/crimson/common/tri_mutex.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+/// shared/exclusive mutual exclusion
+///
+/// similar to seastar::shared_mutex, but instead of two kinds of  waiters,
+/// tri_mutex keeps track of three kinds of them:
+/// - readers
+/// - writers
+/// - exclusive users
+/// and unlike shared_mutex, tri_mutex have two kinds of shared users of lock:
+/// - readers
+/// - writers, which are not mutual-exclusive
```
Comment (athanatos) [inline src/crimson/common/tri_mutex.h:18]: Writers are meant to be exclusive, I'd think.

File: src/crimson/osd/object_context.h [hunk 3]
```diff
@@ -203,14 +203,16 @@ class ObjectContext : public Blocker,
...
     return put_lock_type(RWState::RWREAD);
   }
-  seastar::future<bool> get_recovery_read(bool can_wait = false) {
-    if (bool acquired = rwstate.get_recovery_read();
-        acquired || !can_wait) {
-      return seastar::make_ready_future<bool>(acquired);
+  bool get_recovery_read() {
+    return rwstate.get_recovery_read();
+  }
+  seastar::future<> wait_recovery_read() {
```
Comment (xxhdx1985126) [inline src/crimson/osd/object_context.h:191]: This seems to have been modified by commit "crimson: use tri_mutex for guarding ObjectContext", maybe it should be merged with that commit:)

Comment (tchaikov) [reply src/crimson/osd/object_context.h:191]: but this change is orthogonal to the tri_mutex change. i think they should be different commits.

File: src/crimson/common/tri_mutex.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+/// the exclusive users is like the writers in shared_mutex.
+class tri_mutex {
+public:
+  tri_mutex() = default;
+  // for shared readers
+  seastar::future<> lock_for_read() {
+    if (try_lock_for_read()) {
+      return seastar::make_ready_future<>();
+    }
+    waiters.emplace_back(seastar::promise<>(), type_t::write);
```
Comment (cyx1231st) [inline src/crimson/common/tri_mutex.h:28]: `type_t::read`?

File: src/crimson/common/tri_mutex.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    if (--writers == 0) {
+      wake();
+    }
+  }
+  // for exclusive users
+  seastar::future<> lock_for_excl() {
+    if (try_lock_for_excl()) {
+      return seastar::make_ready_future<>();
+    }
+    waiters.emplace_back(seastar::promise<>(), type_t::write);
```
Comment (cyx1231st) [inline src/crimson/common/tri_mutex.h:73]: `type_t::exclusive`?

**Comment by tchaikov** [comment]
@cyx1231st fixed and repushed. i also moved the implementation into the .cc file.

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2020-09-15_10:24:22-crimson-rados-wip-47311-kefu-distro-basic-smithi

**Comment by tchaikov** [comment]
jenkins test crimson perf

File: src/crimson/common/tri_mutex.h (hunk 2 above)
Comment (rzarzynski) [reply src/crimson/common/tri_mutex.h:18]: I share the same initial confusion – *non-mutual-exclusive writers* are a bit counter-intuitive. How about commenting on that?

File: src/crimson/common/tri_mutex.cc [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+void tri_mutex::wake()
+{
+  assert(!readers && !writers && !exclusively_used);
+  type_t type = type_t::none;
+  while (!waiters.empty()) {
+    auto& waiter = waiters.front();
+    if (type == type_t::exclusive) {
+      break;
```
Comment (rzarzynski) [inline src/crimson/common/tri_mutex.cc:108]: OK, if previous iteration of the loop has encountered a waiter-for-excl, then we shouldn't fulfill any other promise. The execution will be resumed when the exclusive user unlocks.

File: src/crimson/osd/object_context.h [hunk 7]
```diff
@@ -84,33 +85,8 @@ class ObjectContext : public Blocker,
...
-	    return seastar::make_ready_future<seastar::stop_iteration>(
-	      seastar::stop_iteration::no);
-	  });
-	}
-      });
-    });
-  }
-
+  tri_mutex rwlock;
+  bool recovery_read_marker = false;
```
Comment (rzarzynski) [inline src/crimson/osd/object_context.h:99]: Hmm, looks the sole purpose of this flag is to `assert()` in `drop_recovery_read()`. If that's correct, then [`NDEBUG`](https://stackoverflow.com/a/29253284) could be handy. On the other hand, guards aren't elegant...

File: src/crimson/osd/object_context.h [hunk 8]
```diff
@@ -123,8 +123,7 @@ class ObjectContext : public Blocker,
     case RWState::RWNONE:
       return seastar::make_ready_future<>();
     default:
-      ceph_abort_msg("invalid lock type");
```
Comment (rzarzynski) [inline src/crimson/osd/object_context.h:145]: BTW: `ceph_abort_msg()` should be already marked with `[[noreturn]]`. If compiler ignores that, then it's pity.

File: src/crimson/common/tri_mutex.h [hunk 9]
```diff
@@ -35,6 +35,16 @@ class tri_mutex {
 
   bool is_acquired() const;
 
+  /// pass the provided exception to any waiting waiters
+  template<typename Exception>
+  void abort(Exception ex) {
+    while (!waiters.empty()) {
+      auto& waiter = waiters.front();
+      waiter.pr.set_exception(std::make_exception_ptr(ex));
```
Comment (rzarzynski) [inline src/crimson/common/tri_mutex.h:43]: ACK.

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2020-09-15_14:31:19-crimson-rados-wip-kefu-testing-2020-09-15-1950-distro-basic-smithi/

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2020-09-15_14:31:19-crimson-rados-wip-kefu-testing-2020-09-15-1950-distro-basic-smithi/

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Review by rzarzynski — APPROVED** [review]

File: src/crimson/common/tri_mutex.h (hunk 2 above)
Comment (tchaikov) [reply src/crimson/common/tri_mutex.h:18]: agreed. i really need to have a better understanding of pipelined write before explaining this. and this is a bigger topic related to cacheable and rmw operations, AFAICT.

i'd prefer deferring this for a future PR.

File: src/crimson/osd/object_context.h (hunk 7 above)
Comment (tchaikov) [reply src/crimson/osd/object_context.h:99]: no, please see https://github.com/ceph/ceph/pull/37131/files#diff-54370af98a6f12723934c4dff4e882e6R93 .

Comment (rzarzynski) [reply src/crimson/osd/object_context.h:99]: Ah, sure. Since aeb20dd2f5d757d1aa377022977d2277f4eac8ad it's necessary also for release builds.

**Comment by tchaikov** [comment]
@athanatos could you take a final look?

**Comment by tchaikov** [comment]
jenkins test crimson perf

File: src/crimson/common/tri_mutex.h (hunk 2 above)
Comment (athanatos) [reply src/crimson/common/tri_mutex.h:18]: I'd suggest that the fact that this lock design uses reader and writer is entirely and completely independent of the conventional reader/writer usage.  Here, what we mean is that we can pipeline reads, and we can pipeline writes, but we cannot allow a read while writes are in progress or a write while reads are in progress.  Any rmw operation is therefore exclusive.

**Review by athanatos — APPROVED** [review]

File: src/crimson/common/tri_mutex.h (hunk 2 above)
Comment (tchaikov) [reply src/crimson/common/tri_mutex.h:18]: @athanatos thanks Sam. i will incorporate your comment into the code.

Comment (tchaikov) [reply src/crimson/common/tri_mutex.h:18]: #37172

**Pull request merged**